Allow placing a system wide config in /etc/icepapcms (linux) or %systemdrive%/ProgramData/IcePAP (windows)

HEnquist opened the pull request:
At MAX-IV we deploy IcePAPCMS centrally, with separate icepap database configuration for each beamline. This makes sure that any settings are stored in the mysql database, and not by mistake spread out in the home folders of the different people doing the work. 
Earlier we have used a locally modified version of CMS, but we think this functionality would also benefit others.

Modifications in this PR
When starting, we first check if there is a config at /etc/icepapcms/icepapcms.cfg. If it's not there, then use the one in the user home directory.
Use the "--user" argument to force it to use the config from the home dir.
There is also the "--config-file" argument to specify a particular config file.

rhomspuron commented:
Hi @HEnquist,

This is a feature we would like to implement because we have the same problem. The directory /etc/... does not exist on Windows, and the application can be used on Windows so far. Could you find a solution for both (Unix and Windows)? 

Thanks,
Roberto

HEnquist commented:
Ah yes of course, good point! That shouldn't be a problem to handle.
Do you have a suggestion for a suitable path for windows?

HEnquist commented:
> I'm not sure where save this file. As first solution we can implement this command only for Unix, and if it is Windows we have the previous behavior.

That works of course, but would be nicer if we can come up with something that works also on windows. We are looking into it!

> 
> On the other hand, if we configure to read from /etc/xxx, the user does not have access to change the parameters, in this case the GUI can crash if we change something on the preference dialog. One possible solution is to do not allow to open the configuration dialog when it is using the system configuration. This solution has a problem, because the user can't change where to load/save firmware, snapshots, etc. In our case we would like to have a configuration where the user can't change this parameters and should be common for all the users.
>
It doesn't crash, if you try to save to the system wide config you get a message: "You must run IcePAPCMS as superuser to change the configuration parameters, or use a local config file."
 
> Another point, if we have the configuration on the system, I'm not sure if it is a good idea to allow run the system in user mode, because if you have a wrong configuration on the user folder it can generate problems. Here there are two types of configuration IMHO:
> 
>     1. database, ldap, network (should be on the system, administrator only can change)
> 
>     2. folders configuration: snapshots, firmware, etc. (the user can change them)

We have changed only what happens with the config, the rest like snapshots etc still go to the home dir.

We made the user mode with advanced users in mind. It's overlapping with the functionality provided by the -c option, but it's less typing. I don't really see a problem with allowing it also when a system-wide config exists, since I expect it to be used only by the hardware team who know what they are doing.


> 
> 
> As first implementation, we can do:
> 
>     1. If there is configuration file on the system /etc/xxx: the application block the configuration dialog, it allows only to read, but I don't know how to use it to change parameters.

The current implementation doesn't block, but it gives a warning if trying to save to the system config.
> 
>     2. if there is not file or the system is windows, we have the current behavior.
>        In this case the option --user is not implemented.
>        I think that the option --config-file is a good solution, it allows to bypass the system configuration (/etc/xxx) and it is for experts in the case that they don't know the root password to change the system configuration.
>        What do you think?
> 




> For the system path maybe it is better to use /etc/icepap/xxx or /etc/icepapsuite/xxx in this case we can set configuration for all programs: icepap library, icepapcms, icepaposc, etc.

Good point, I would vote for simply /etc/icepap then.

rhomspuron commented:
> That works of course, but would be nicer if we can come up with something that works also on windows. We are looking into it!

Totally agree

> It doesn't crash, if you try to save to the system wide config you get a message: "You must run IcePAPCMS as superuser to change the configuration parameters, or use a local config file."

Great, I'd take out the "use a local configuration file", the expert user knows how to do it.

>We made the user mode with advanced users in mind. It's overlapping with the functionality provided by the -c option, but it's less typing. I don't really see a problem with allowing it also when a system-wide config exists, since I expect it to be used only by the hardware team who know what they are doing.


In our case, the control group configures the database and that's why we run only the IcepapCMS with one user (operator), the other uses on the beamline have no configuration, we don't like it. It is true that now they can change it, that is why we would like this feature.  IMHO if we configure from the system we should not allow to have configuration in the user, in our case I can not identify in the beamline more than one configuration, in your case do you need more than one configuration in the beamlines? In labs it is good to have more than one configuration, but in this case the installation is different for us, the installation is per user and therefore it can be interesting to be able to change quickly with the -c option.

Anyway I don't see any problem in implementing the -u and -c option. I was saying it to avoid possible problems like the system configuration does not allow to see all icepaps of the network and the user configuration does. In our case when we install it we will move the user configuration to the system, that's why I don't see any problems.

HEnquist commented:
We believe we have found a nice solution for this on Windows. We'll update the PR very soon.

> 
> Great, I'd take out the "use a local configuration file", the expert user knows how to do it.
> 

Sure, the experts already know, and we should not give the non-experts any ideas :)  

> In our case, the control group configures the database and that's why we run only the IcepapCMS with one user (operator), the other uses on the beamline have no configuration, we don't like it. It is true that now they can change it, that is why we would like this feature. IMHO if we configure from the system we should not allow to have configuration in the user, in our case I can not identify in the beamline more than one configuration, in your case do you need more than one configuration in the beamlines? In labs it is good to have more than one configuration, but in this case the installation is different for us, the installation is per user and therefore it can be interesting to be able to change quickly with the -c option.
> 

We also have a single database and config per beamline, but we usually work using our own accounts. When setting up a new beamline, the software team starts by setting up the beamline database, and deploying the matching icepapcms system config. Then every user automatically gets the right config from the system, so whoever works on it always uses the one correct database. Or to be precise, that's how it used to work, before we upgraded to icepapcms from upstream. Then the old patches didn't work anymore, and lost the system config functionality. This has caused quite some confusion..

In the controls labs it's a little different. We have all the labs configured almost like one big beamline. When working there, then the "-u" and "-c" modes become handy. Most machines on the lab "beamline" will have a system config, that may or may not be very useful.

HEnquist commented:
This works for us now on both linux and windows. Can you take a look?

Another thing, the Preferences dialog, there is a filed for changing the config path. Is we actually allow changing that path there? , 

This path also gets saved in the config file itself, with the result that the config file has a field pointing at itself..

HEnquist commented:
The systemwide config is now placed at:
- Windows: `%systemdrive%/ProgramData/IcePAP`
- Linux: `/etc/icepapcms`

I made a separate issue for the config path setting: https://github.com/ALBA-Synchrotron/IcepapCMS/issues/45

HEnquist commented on icepapcms/gui/dialogpreferences.py:
```diff
@@ -58,16 +59,27 @@ def __init__(self, parent):
-                self, "Preferences", "Check configuration parameters")
+            MessageDialogs.showWarningMessage(self, "Preferences", 
+                "You must run IcePAPCMS as superuser to change the configuration"
+                " parameters, or use a local config file.")
```
done!

HEnquist commented on icepapcms/gui/dialogpreferences.py:
```diff
@@ -58,16 +59,27 @@ def __init__(self, parent):
         self._config = ConfigManager()
         self.fillConfig()
         self.ui.listWidget.item(0).setSelected(True)
+        self.setWindowTitle("Preferences (%s)" % self._config.config_filename)
```
done!

HEnquist commented:
The config path is now removed from the config file. The preferences dialog looks like this:
![prefs](https://user-images.githubusercontent.com/6504678/130599783-98f54ea2-2ac0-4cc3-8aa2-ae1cd205cb78.png)

I also added an item in the context meny to open an icepap console with the right hostname already fille in:
![console_context](https://user-images.githubusercontent.com/6504678/130600019-a9d60c75-1ea1-4267-8adf-8ddee7ff5e4f.png)

Then it opens with the host and port pre-filled:
![console_opened](https://user-images.githubusercontent.com/6504678/130600068-27f769b4-7ada-49ed-a453-ebd1c6387c5c.png)

Finally I also fixed a few other bugs, like that creating a new sqlite database file failed.

HEnquist commented:
Is there any remaining points I need to take care of now?

cpascual commented:
> The systemwide config is now placed at:
> 
>     * Windows: `%systemdrive%/ProgramData/IcePAP`
>     * Linux: `/etc/icepapcms`

Hi, just a comment in case you are interested: in taurus we recently did something similar but we chose to follow the xdg standard for the default config dirs on posix systems and also a different (IMHO, more generic) location for windows. You can see more details in https://gitlab.com/taurus-org/taurus/-/merge_requests/1205

HEnquist commented:
Ah yes thanks, we should definitely use `%PROGRAMDATA%` instead of`%SYSTEMDRIVE%/ProgramData`. I'll change that.

As for following xdg or not, I don't really have any preference. @rhomspuron ?

On Windows, using `%APPDATA%` for the user config is the "right" way. But I see one advantage of `~/.icepapcms/configs/icepapcms.conf` (or xdg `~/.config/icepapcms/...`).  If the home folder lives on the network, and the same folder is used for both windows and linux machines, then the same config file will be used no matter which OS the user ended up using today. Then I don't know how useful that is in practice :)

rhomspuron commented:
Hi @HEnquist 
> As for following xdg or not, I don't really have any preference. @rhomspuron ?

I don't have any preference either

> On Windows, using `%APPDATA%` for the user config is the "right" way. But I see one advantage of `~/.icepapcms/configs/icepapcms.conf` (or xdg `~/.config/icepapcms/...`). If the home folder lives on the network, and the same folder is used for both windows and linux machines, then the same config file will be used no matter which OS the user ended up using today. Then I don't know how useful that is in practice :)

Good point, at Alba we do not install it on Windows, I don't know if it is a case of use. 

I will review the changes. 

Thanks!

rhomspuron commented:
Hi @HEnquist 

It is great, thanks!!!

Roberto
